File: quick_sort.py
# In-place Quick Sort (pivot: random item), timing, and plotting for multiple input files
import time
import random

def quick_sort(arr, low, high):
    while low < high:
        pi = partition(arr, low, high)
        # Tail call elimination: sort smaller part first
        if pi - low < high - pi:
            quick_sort(arr, low, pi - 1)
            low = pi + 1
        else:
            quick_sort(arr, pi + 1, high)
            high = pi - 1

def partition(arr, low, high):
    # Select a random pivot and swap with high
    pivot_index = random.randint(low, high)
    arr[pivot_index], arr[high] = arr[high], arr[pivot_index]
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def main(inputs_dict = None):
    print("Running Quick Sort...")
    if inputs_dict is None:
        print("No input provided.")
        return
    else:
        # Dynamic input handling
        results = {
            'random': {'sizes': [], 'times': []},
            'sorted': {'sizes': [], 'times': []},
            'reverse_sorted': {'sizes': [], 'times': []}
        }
        for key in ['random', 'sorted', 'reverse_sorted']:
            for n, arr in inputs_dict.get(key, {}).items():
                times = []
                num_runs = 3 if n > 10000 else 5
                for _ in range(num_runs):
                    arr_copy = arr.copy()
                    start = time.perf_counter()
                    quick_sort(arr_copy, 0, len(arr_copy) - 1)
                    end = time.perf_counter()
                    times.append(end - start)
                avg_time = sum(times) / len(times)
                results[key]['sizes'].append(n)
                results[key]['times'].append(avg_time)
                print(f"{key.capitalize()} Input size: {n}, Average time over {num_runs} runs: {avg_time:.6f} seconds")
        print("******* Quick Sort Finished *******")
        return results

def get_results(inputs_dict = None):
    return main(inputs_dict)

if __name__ == "__main__":
    main()

File: quick_sort.cpp
// Quick Sort, timing, and input generation (C++)
#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <algorithm>
#include <random>

int partition(std::vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; ++j) {
        if (arr[j] <= pivot) {
            ++i;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}
void quick_sort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quick_sort(arr, low, pi - 1);
        quick_sort(arr, pi + 1, high);
    }
}

std::vector<int> generate_input(size_t size, const std::string& type) {
    std::vector<int> arr(size);
    if (type == "random") {
        std::mt19937 rng(42);
        std::uniform_int_distribution<int> dist(0, size * 10);
        for (size_t i = 0; i < size; ++i) arr[i] = dist(rng);
    } else if (type == "sorted") {
        for (size_t i = 0; i < size; ++i) arr[i] = i;
    } else if (type == "reverse_sorted") {
        for (size_t i = 0; i < size; ++i) arr[i] = size - i;
    }
    return arr;
}

int main() {
    std::cout << "Running Quick Sort..." << std::endl;
    std::vector<size_t> sizes = {1000,2000,3000,4000,5000,10000,20000,40000,50000,60000,80000,90000,100000};
    std::vector<std::string> types = {"random", "sorted", "reverse_sorted"};
    for (const auto& type : types) {
        for (auto size : sizes) {
            std::vector<int> arr = generate_input(size, type);
            int num_runs = (size > 10000) ? 3 : 5;
            double total_time = 0.0;
            for (int run = 0; run < num_runs; ++run) {
                std::vector<int> arr_copy = arr;
                auto start = std::chrono::high_resolution_clock::now();
                quick_sort(arr_copy, 0, arr_copy.size() - 1);
                auto end = std::chrono::high_resolution_clock::now();
                std::chrono::duration<double> elapsed = end - start;
                total_time += elapsed.count();
            }
            double avg_time = total_time / num_runs;
            std::cout << "Type: " << type << " | Size: " << size << ", Avg time: " << avg_time << " seconds" << std::endl;
        }
    }
    std::cout << "******* Quick Sort Finished *******" << std::endl;
    return 0;
}
